lib/ts/controllers/s-uploader.ts

```diff
@@ -0,0 +1,171 @@
+namespace Stacks {
+    interface FilePreview {
+        data?: string | ArrayBuffer;
+        name: string;
+        type: string;
+    };
+
+    export class UploaderController extends Stacks.StacksController {
+        static targets = ["input", "previews", "uploader"];
+        private inputTarget!: HTMLInputElement;
+        private previewsTarget!: HTMLElement;
+        private uploaderTarget!: HTMLElement;
+
+        private boundDragEnter!: any;
+        private boundDragLeave!: any;
+
+        private static readonly FILE_DISPLAY_LIMIT = 10;
+        private static readonly MAX_FILE_SIZE = 1024 * 1024 * 10; // 10 MB
+
+        connect() {
+            super.connect();
+            this.boundDragEnter = this.handleUploaderActive.bind(this, true);
+            this.boundDragLeave = this.handleUploaderActive.bind(this, false);
+
+            this.inputTarget.addEventListener("dragenter", this.boundDragEnter);
+            this.inputTarget.addEventListener("dragleave", this.boundDragLeave);
+        }
+
+        disconnect() {
+            this.inputTarget.removeEventListener("dragenter", this.boundDragEnter);
+            this.inputTarget.removeEventListener("dragleave", this.boundDragLeave);
+            super.disconnect();
+        }
+
+        /**
+         * Handles rendering the file preview state on input change
+         */
+        handleInput() {
```

Comment:
I'm sorta back and forth on "expose this for use in `data-action`" and "just wire it up always, don't make the user do it". In pure stimulus, we'd prefer the former, but I'm not really super into "pure stimulus". This is fine as-is, but I did want to bring it up to prompt the discussion.

Reply:
Admittedly, I was leaning heavily on prior art (I can't recall which component I referenced most), but I think stepping away from "pure stimulus" is generally a good call.

It probably does make more sense to just attach this function directly to the target input but I'm good to leave it as-is for now.